cni/network/network.go

```diff
@@ -839,3 +868,102 @@ func (plugin *netPlugin) Update(args *cniSkel.CmdArgs) error {
 
 	return nil
 }
+
+// Temporary function to determine whether we need to disable SNAT due to NMAgent support
+func determineSnatOnHost(nwCfg *cni.NetworkConfig) error {
+	var (
+		snatOnHostConfig      snatOnHostConfiguration
+		retrieveSnatConfigErr error
+		jsonFile              *os.File
+	)
+
+	snatOnHostConfigFile := snatOnHostConfigFileName + jsonFileExtension
+
+	// Check if we've already retrieved NMAgent version and determined whether to disable snat on host
+	if jsonFile, retrieveSnatConfigErr = os.Open(snatOnHostConfigFile); retrieveSnatConfigErr == nil {
+		bytes, _ := ioutil.ReadAll(jsonFile)
+		retrieveSnatConfigErr = json.Unmarshal(bytes, &snatOnHostConfig)
+		jsonFile.Close()
+	}
+
+	// If we weren't able to retrieve snatOnHostConfiguration, query NMAgent
+	if retrieveSnatConfigErr != nil {
+		var resp *http.Response
+		resp, retrieveSnatConfigErr = http.Get(nmAgentVersionURL)
+		if retrieveSnatConfigErr == nil {
+			defer resp.Body.Close()
+
+			if resp.StatusCode == http.StatusOK {
+				var bodyBytes []byte
+				// if the list of APIs (strings) contains the newNMAgentAPI we will disable snat on host
+				if bodyBytes, retrieveSnatConfigErr = ioutil.ReadAll(resp.Body); retrieveSnatConfigErr == nil {
+					if strings.Contains(string(bodyBytes), newNMAgentAPI) {
+						snatOnHostConfig.DisableSnatOnHost = true
+					}
+
+					if retrieveSnatConfigErr = determineSnatOnHostLock(); retrieveSnatConfigErr == nil {
+						ioutil.WriteFile(snatOnHostConfigFile, []byte(fmt.Sprintf("%+v", snatOnHostConfig)), 0644)
+						os.Remove(snatOnHostConfigFileName + lockFileExtension)
+					}
+				}
+			} else {
+				retrieveSnatConfigErr = fmt.Errorf("nmagent request status code %d", resp.StatusCode)
+			}
+		}
+	}
+
+	// Log and return the error when we fail to determine whether to disable snat on host
+	if retrieveSnatConfigErr != nil {
+		log.Errorf("[cni-net] failed to determine whether to disable snat on host with error %v",
+			retrieveSnatConfigErr)
+		return retrieveSnatConfigErr
+	}
+
+	nwCfg.EnableSnatOnHost = !snatOnHostConfig.DisableSnatOnHost
+
+	log.Printf("[cni-net] EnableSnatOnHost set to %t", nwCfg.EnableSnatOnHost)
+
+	return nil
+}
+
+// determineSnatOnHostLock acquires lock for write access to disableSnatOnHost.json
+func determineSnatOnHostLock() error {
+	var (
+		lockFile *os.File
+		lockName = snatOnHostConfigFileName + lockFileExtension
+		err      error
+
+		// Try to acquire the lock file.
+		lockRetryCount uint
+		modTimeCur     time.Time
+		modTimePrev    time.Time
+	)
+
+	for lockRetryCount < maxLockRetries {
+		lockFile, err = os.OpenFile(lockName, os.O_CREATE|os.O_EXCL|os.O_RDWR, os.FileMode(0664)+os.FileMode(os.ModeExclusive))
```

Comment:
can it move this to varaible os.FileMode(0664)+os.FileMode(os.ModeExclusive) -> lockperm so that it will be easier to read